soc/xtensa/intel_adsp/tools/cavstool.py

```diff
@@ -659,30 +662,59 @@ async def _main(server):
         if dsp.HIPCIDA & 0x80000000:
             dsp.HIPCIDA = 1<<31 # must ACK any DONE interrupts that arrive!
 
+        if server:
+            # Check if the client connection is alive.
+            if not is_connection_alive(server):
+                lock.acquire()
+                start_output = False
+                lock.release()
+
 class adsp_request_handler(socketserver.BaseRequestHandler):
     """
     The request handler class for control the actions of server.
     """
 
-    def receive_fw(self, filename):
+    def receive_fw(self):
+        header_sz = 80
+
+        log.info("Receiving...")
+        # The first contain the header and the front
+        # data of the firmware
+        d = self.request.recv(BUF_SIZE)
+
+        # Unpacked the header data
+        # Include size(4), filename(42), MD5(32) and pad(2)
+        header = d[:header_sz]
+        total = d[header_sz:]
+        s = struct.Struct(PACKET_HEADER_FORMAT_FW)
+        fsize, fname, md5_tx_b, pad = s.unpack(header)
+        log.info(f'size:{fsize}, filename:{fname}, MD5:{md5_tx_b}, pad:{pad})')
+
+        # Receive the remain part of the firmware. We only
+        # receive the specified amount of bytes.
+        while len(total) < fsize:
+            data = self.request.recv(min(BUF_SIZE, fsize - len(total)))
```

Comment:
When I asked to remove the `if not data; break` it was only temporarily for testing purposes. You still need to check for disconnections otherwise you can trigger an infinite loop. You need this:

```diff
--- a/soc/xtensa/intel_adsp/tools/cavstool.py
+++ b/soc/xtensa/intel_adsp/tools/cavstool.py
@@ -710,6 +710,8 @@ class adsp_request_handler(socketserver.BaseRequestHandler):
         # receive the specified amount of bytes.
         while len(total) < fsize:
             data = self.request.recv(min(BUF_SIZE, fsize - len(total)))
+            if not data:
+                raise EOFError("truncated firmware file")
             total += data
 

```

Here's how to test or reproduce the infinite loop:

```diff
--- a/soc/xtensa/intel_adsp/tools/cavstool_client.py
+++ b/soc/xtensa/intel_adsp/tools/cavstool_client.py
@@ -73,7 +73,15 @@ class cavstool_client():
             log.info(f'Sending...')
 
             total = self.sock.send(header_data)
-            total += self.sock.sendfile(f)
+
+            fw = f.read()
+            total += len(fw)
+
+            if False:
+                self.sock.send(fw) # Works fine
+            else:  # truncationg hangs forever; then infinite loop on the test device after
+                # interrupting cavstool_client.py with Ctrl-C
+                self.sock.send(fw[:-5])
 
             log.info(f"Done Sending ({total}).")

```

Reply:
That's fantastic! Nice catch! Thank you!